File: src/main/java/nl/jhvh/java/geometry/model/twodimensional/TwoDimensional.java
package nl.jhvh.java.geometry.model.twodimensional;

import nl.jhvh.java.geometry.model.Dimensional;
import nl.jhvh.java.geometry.model.Geometry;

import static nl.jhvh.java.geometry.model.Dimensional.twoDimensional;

interface TwoDimensional extends Geometry {

    /** Circumference of the {@link TwoDimensional} geometry in meters */
    double getCircumference();

    /** Area of a {@link TwoDimensional} geometry in square meters */
    double getArea();

    @Override
    default Dimensional getDimensional() {
        return twoDimensional;
    }
}

File: src/main/kotlin/nl/jhvh/kotlin/geometry/model/twodimensional/TwoDimensional.kt
package nl.jhvh.kotlin.geometry.model.twodimensional

import nl.jhvh.kotlin.geometry.model.Geometry

interface TwoDimensional: Geometry, Comparable<TwoDimensional> {

    /** Circumference of the [TwoDimensional] geometry in meters */
    val circumference: Double

    /** Area of a [TwoDimensional] geometry in square meters */
    val area: Double

    /** @return [Double] The sum of the [TwoDimensional.area]s */
    operator fun <T: TwoDimensional, S: TwoDimensional>T.plus(other: S): Double

    /** @return [Double] The subtraction of the [TwoDimensional.area]s */
    operator fun <T: TwoDimensional, S: TwoDimensional>T.minus(other: S): Double

}

operator fun <T: TwoDimensional, S: TwoDimensional> T.plus(other: S): Double = this.plus(other)
operator fun <T: TwoDimensional, S: TwoDimensional> T.minus(other: S): Double = this.minus(other)
